storage/src/lib.rs

```diff
@@ -410,22 +410,70 @@ impl Storage {
     /// - Ok of Some entry - if found in loose index
     /// - Ok of None - if requested height is low enough to not be in the index already
     /// - Err of `BlockHeightNotFound` - if requested height is too large and out of index bounds
+    /// NOTE: this function does NOT return EBB, even if it's a part of the loose index
     pub fn get_from_loose_index(
         &self,
         height: ChainDifficulty,
     ) -> Result<Option<&LooseChainHeightEntry>> {
-        if let Some(ref idx_tip) = self.chain_height_idx.loose_idx.first() {
-            let find_diff = u64::from(height);
-            let tip_diff = u64::from(idx_tip.difficulty);
-            if find_diff > tip_diff {
-                return Err(Error::BlockHeightNotFound(find_diff));
+        match self.chain_height_idx.loose_idx.first() {
+            None => {
+                // No loose index is available at the moment at all
+                return Ok(None);
             }
-            let idx = (tip_diff - find_diff) as usize;
-            if self.chain_height_idx.loose_idx.len() > idx {
-                return Ok(Some(&(self.chain_height_idx.loose_idx[idx])));
+            Some(ref idx_tip) => {
+                // idx_tip is the NEWEST loose block
+                let find_diff = u64::from(height);
+                let tip_diff = u64::from(idx_tip.difficulty);
+                if find_diff > tip_diff {
+                    return Err(Error::BlockHeightNotFound(find_diff));
+                }
+                let idx = (tip_diff - find_diff) as usize;
+                let loose_index_len = self.chain_height_idx.loose_idx.len();
+                if loose_index_len <= idx {
+                    // We are searching deeper in history than the loose index
+                    return Ok(None);
+                }
+                let entry = &self.chain_height_idx.loose_idx[idx];
+                if !entry.date.is_boundary() {
+                    return Ok(Some(entry));
+                }
+                debug!("Search by height in loose index found EBB. Proceeding with caution.");
+                if (loose_index_len - idx) <= 1 {
+                    debug!("EBB is the oldest available loose entry. Returning nothing, to continue search in historical data.");
+                    return Ok(None);
+                }
+                // There are loose blocks older than found EBB, trying next one
```

Comment:
Are we running into a situation where our loose index contains multiple epochs and the EBB isn't the last block in the vec, or is this just in case?